### ISSUE TITLE
just one MQTT message in one TCP-IP packet

### ISSUE THREAD

**Issue by radoslavv** [eclipse/mosquitto#970]
Hi,

Got new requirement for consideration.

Could you implement command line argument that will prevent bundling of several MQTT messages into one TCP-IP packet sent to a connected client?
Actually I am developing client based on Microchip's TCP-IP lite stack on PIC18F class uP that does have resources limits preventing processing more than one MQTT message per TCP-IP packet (mostly due to RAM & stack limitation).
As I am not able to process packets right way on uP, could you please allow on Mosquitto sending not more than one MQTT message per TCP-IP packet, please? Ideally defined per IP, as not all clients/servers may have this limitation (like a configuration parameter). This will help very much.
I suppose this is not the first case of low cost uP implementation with such an need.
Thank you.

Rado
(actually I am using Windows version, but planning to switch to Ubuntu later)

**Comment by ralight** [comment]
Could you try setting `set_tcp_nodelay true` in your configuration file? Mosquitto doesn't handle packets in any special way other than to try to minimise writes to help ensure that small messages don't end up on multiple packets.

**Comment by karlp** [comment]
You really should evaluate your mqtt client library again, if it can't handle this.  People have been implementing mqtt clients on very constrained devices for _years_ without needing this.

**Comment by radoslavv** [comment]
> You really should evaluate your mqtt client library again, if it can't handle this. People have been implementing mqtt clients on very constrained devices for _years_ without needing this.

Anybody showing me such an implementation on pic18f family with 3kB RAM running TCP-IP lite stack with paho.mqtt.embedded-c is welcome.

**Comment by radoslavv** [comment]
> Could you try setting `set_tcp_nodelay true` in your configuration file? Mosquitto doesn't handle packets in any special way other than to try to minimise writes to help ensure that small messages don't end up on multiple packets.

does not helped, still multiple messages in one incoming tcp-ip packet (QoS1 PUBACKs) when publishing several messages quickly one after another

**Comment by ralight** [comment]
In that case, you may wish to try restricting yourself to processing a single message at once. In other words, only send your second publish after the first puback has been received.

**Comment by karlp** [comment]
bear in mind, you're asking for broker changes to try and work around standard tcp behaviour of sockets, to work aroudn a problem in your (broken, or improperly configured/used) tcp stack and your (broken or improperly configured/used) mqtt client library.

Even if you could get broker changes in _this_ broker, you're going to have to go through all the same problems whenever the broker you're connecting to is switched out for another (compliant) broker for management or scalability reasons.  

QoS 1 or 2 on a highly constrained device is likely also going to be problematic for you.

**Comment by ralight** [comment]
I'm afraid my inclination is to close this as "won't fix". From what I've seen, the only realistic way to implement this would be set/unset TCP_NODELAY every time an MQTT message gets sent, which feels like a pretty crude hammer to fix it with.

**Comment by radoslavv** [comment]
> In that case, you may wish to try restricting yourself to processing a single message at once. In other words, only send your second publish after the first puback has been received.

Normally I do it this way.
I found PINGREQ is sent independently of this sequential PUBLISH approach, so sometimes I am receiving packet with PINGRESP+PUBACK.

However I am using also SUBSCRIBE to some topics, and I can still receive PUBLISHes from broker + PUBACKs in one packet.
I have some idea how to reimplement MQTT client (paho.mqtt.embedded-c).

Closing ticket.
